File: packages/agentaid-ts/src/otel/exporter.ts
import { ExportResult, ExportResultCode } from "@opentelemetry/core";
import type { ReadableSpan, SpanExporter } from "@opentelemetry/sdk-trace-base";

interface SerializedSpan {
  trace_id: string;
  span_id: string;
  parent_span_id: string | null;
  name: string;
  kind: string;
  start_time_unix_nano: number;
  end_time_unix_nano: number;
  attributes: Record<string, unknown>;
  events: Array<{ name: string; timestamp_unix_nano: number; attributes: Record<string, unknown> }>;
  status: { code: string; description: string };
}

function _serialize(span: ReadableSpan): SerializedSpan {
  const ctx = span.spanContext();
  return {
    trace_id: ctx.traceId,
    span_id: ctx.spanId,
    parent_span_id: span.parentSpanId ?? null,
    name: span.name,
    kind: String(span.kind),
    start_time_unix_nano: span.startTime[0] * 1e9 + span.startTime[1],
    end_time_unix_nano: span.endTime[0] * 1e9 + span.endTime[1],
    attributes: { ...span.attributes },
    events: span.events.map(e => ({
      name: e.name,
      timestamp_unix_nano: e.time[0] * 1e9 + e.time[1],
      attributes: { ...(e.attributes ?? {}) },
    })),
    status: { code: String(span.status.code), description: span.status.message ?? "" },
  };
}

export class AgentAidSpanExporter implements SpanExporter {
  constructor(public endpoint = process.env.AGENTAID_ENDPOINT ?? "http://localhost:8000/ingest") {}

  export(spans: ReadableSpan[], cb: (result: ExportResult) => void): void {
    const payload = { spans: spans.map(_serialize) };
    fetch(this.endpoint, {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(payload),
    })
      .then(() => cb({ code: ExportResultCode.SUCCESS }))
      .catch((err) => cb({ code: ExportResultCode.FAILED, error: err as Error }));
  }

  async shutdown(): Promise<void> { /* fetch is fire-and-forget */ }
}

File: packages/agentaid-py/src/agentaid/otel/exporter.py
from __future__ import annotations

import logging
from collections.abc import Sequence

import httpx
from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult
from opentelemetry.trace.span import format_span_id, format_trace_id

log = logging.getLogger(__name__)


def _serialize_span(span: ReadableSpan) -> dict:
    ctx = span.get_span_context()
    parent = span.parent
    return {
        "trace_id": format_trace_id(ctx.trace_id),
        "span_id": format_span_id(ctx.span_id),
        "parent_span_id": format_span_id(parent.span_id) if parent else None,
        "name": span.name,
        "kind": str(span.kind),
        "start_time_unix_nano": span.start_time,
        "end_time_unix_nano": span.end_time,
        "attributes": dict(span.attributes or {}),
        "events": [
            {"name": e.name, "timestamp_unix_nano": e.timestamp,
             "attributes": dict(e.attributes or {})}
            for e in (span.events or [])
        ],
        "status": {"code": span.status.status_code.name,
                   "description": span.status.description or ""},
    }


class AgentAidSpanExporter(SpanExporter):
    """OTel exporter that POSTs serialized spans to the AgentAid server.

    BatchSpanProcessor calls export() from a dedicated worker thread, so a
    synchronous HTTP client is correct here — it avoids the event-loop
    coordination issues that an async client would introduce when the host
    process is itself running asyncio (and would otherwise cause spans to
    be lost on process exit).
    """

    def __init__(
        self, endpoint: str = "http://localhost:8000/ingest", timeout: float = 5.0
    ) -> None:
        self.endpoint = endpoint
        self.timeout = timeout
        self._client = httpx.Client(timeout=timeout)

    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        if not spans:
            return SpanExportResult.SUCCESS
        payload = {"spans": [_serialize_span(s) for s in spans]}
        try:
            self._client.post(self.endpoint, json=payload)
            return SpanExportResult.SUCCESS
        except Exception:
            log.warning("agentaid exporter failed", exc_info=True)
            return SpanExportResult.FAILURE

    def shutdown(self) -> None:
        try:
            self._client.close()
        except Exception:
            pass
